### PR TITLE
swf: treat empty gradients as solid black color

### PR DESCRIPTION
**Pull request by adrian17**
Fixes https://github.com/ruffle-rs/ruffle/issues/3365 , https://github.com/ruffle-rs/ruffle/issues/8156 .

### CODE REVIEW

**Review by Aaron1011 — APPROVED** [review]

File: swf/src/read.rs [hunk 1]
```diff
@@ -1576,16 +1576,34 @@ impl<'a> Reader<'a> {
...
+            0x12 => {
+                if let Some(gradient) = self.read_gradient(shape_version)? {
+                    FillStyle::RadialGradient(gradient)
+                } else {
+                    FillStyle::Color(Color::BLACK)
+                }
+            }
 
-            0x13 => FillStyle::FocalGradient {
-                // SWF19 says focal gradients are only allowed in SWFv8+ and DefineShape4,
```
Comment (relrelb) [inline swf/src/read.rs:1584]: Can we keep this comment?

Comment (adrian17) [reply swf/src/read.rs:1584]: Oops sorry, merged before reading this. Will update this later, sure.

File: swf/src/read.rs [hunk 2]
```diff
@@ -1576,16 +1576,34 @@ impl<'a> Reader<'a> {
                 FillStyle::Color(color)
             }
 
-            0x10 => FillStyle::LinearGradient(self.read_gradient(shape_version)?),
+            0x10 => {
```
Comment (relrelb) [inline swf/src/read.rs:1579]: In order to avoid code duplication, it might be worth to unify all 3 gradient cases into a single one using `0x10..=0x13`.

**Comment by relrelb** [comment]
Actually, I think the actual root cause is a bug in the WebGL backend, and is not really related SWF parsing. The following lines seem buggy: https://github.com/ruffle-rs/ruffle/blob/52f773643609b82fcfa442c26f60b43388f28cf4/render/webgl/src/lib.rs#L1403-L1406

It effectively fills the rest `MAX_GRADIENT_COLORS` with the last ratio and color, but that obviously fails when there are no colors. Maybe we can revert this PR and instead just skip that `for` loop in case `num_colors == 0`?

**Comment by adrian17** [comment]
When we talked about it on Discord earlier.
`num_colors == 0` is a malformed SWF, it shouldn't happen at all. We decided to prevent this at the source rather than manually handle this in every render backend. Ideally we'd love some kind of `NonEmptyVec` but for now there's just an assumption that the Gradient has >0 colors, like the spec says. You can add a panic in renderers if you want to be extra safe?

**Comment by relrelb** [comment]
> We talked about it on Discord earlier. `num_colors == 0` is a malformed SWF, spec-wise it shouldn't happen at all. We decided to prevent this at the source rather than manually handle this in every render backend. Ideally we'd love some kind of `NonEmptyVec` but for now there's just an assumption that the Gradient has >0 colors, like the spec says. You can add a panic in renderers if you want to be extra safe?

`num_colors == 0` is indeed disallowed by spec. But from a perspective of crates that use `swf` directly, it's unexpected that `read_fill_style()` can lie about the returned fill style type. In general I think we should avoid Ruffle-specific constraints in `swf` as much as possible. So I still prefer fixing this specifically in WebGL instead. The fix would also be simpler that way.

**Pull request merged**